### ISSUE TITLE
recovery: offload large files to cloud storage

### ISSUE THREAD

**Issue by jgraettinger** [estuary/flow#190]
`recoverylog.Recorder` observes file operations and encodes them into a recovery log journal. It tracks the still-live file manifest and produces "hints" which allow a reader to efficiently read those files.

Frequently (as with RocksDB) a Recorder is observing a mix of small files which are frequently written an synced (for example, the RocksDB `MANIFEST` file) with much larger files which are opened, monotonically written from offset 0, and then closed without ever having been sync'd (for example, most SST files).

There are two performance issues with the current implementation:
 * During cold-start recovery playback, the player must frequently read through lots of recovery-log content it doesn't care about (for example, files known to already be deleted according to playback hints) in order to read interleaved files it _does_ care about (the `MANIFEST`). This increases recovery time.
 * These SST files can be large, and exhibit Rocks write amplification. By virtue of being written to journals, they're replicated across AZ's, they increase load on brokers, and incur network transfer costs.

Proposal:

We can apply a heuristic within the Recorder, which will lazily write file contents to the recovery log _only if_ the file is a) sync'd prior to close or b) written at a non-contiguous append offset. If either a) or b) occur, we write the file to the log as we would today.

If the file _is_ opened / written contiguously / never synced, then we would instead defer any writes of the file to the recovery log.
Upon the file close, we'd
 - Upload the entire file directly to cloud storage keyed on the file's FNode number, and
 - Write an indirection to that file in the recovery log.
 
This keeps the log itself lighter-weight, and focused on encoding manifest changes in between sync barriers. Readers are able to efficiently skip over files they don't care about, and are _also_ able to pull down files they _do_ care about in parallel.

It makes pruning more complicated, but still tractable: we can merge outstanding FSMHints to identify all FNodes which may still exist in any set of hints, and leverage the fact that files are keyed on FNode to efficiently list and remove files which are no longer needed.
